pkg/volume/portworx/portworx_util.go

```diff
@@ -0,0 +1,276 @@
+	replLabel             = "repl"
+	cosLabel              = "io_priority"
+	sharedLabel           = "shared"
+	fsLabel               = "fs"
+	snapshotIntervalLabel = "snapshot_interval"
+	aggregationLabel      = "aggregation_level"
+	blocksizeLabel        = "block_size"
+)
+
+type PortworxVolumeUtil struct{
+	portworxClient *osdclient.Client
+}
+
+func (util *PortworxVolumeUtil) DeleteVolume(d *portworxVolumeDeleter) error {
+	hostName := d.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return err
+	}
+
+	err = client.Delete(d.volumeID)
+	if err != nil {
+		glog.Infof("Error in Volume Delete for (%v): %v", d.volName)
+		return err
+	}
+	return nil
+}
+
+// CreateVolume creates a Portworx volume.
+// Returns: volumeID, volumeSize, labels, error
+func (util *PortworxVolumeUtil) CreateVolume(p *portworxVolumeProvisioner) (string, int, map[string]string, error) {
+	hostName := p.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return "", 0, nil, err
+	}
+
+	capacity := p.options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	// Portworx Volumes are specified in GB
+	requestGB := int(volume.RoundUpSize(capacity.Value(), 1024*1024*1024))
+
+	var labels map[string]string
+	if p.options.CloudTags == nil {
+		labels = make(map[string]string)
+	} else {
+		labels = *p.options.CloudTags
+	}
+
+	spec := osdapi.VolumeSpec{
+		Size: uint64(requestGB*1024*1024*1024),
+		// Default format is btrfs
+		Format: osdapi.FSType_FS_TYPE_BTRFS,
+	}
+
+	for k, v := range p.options.Parameters {
+		switch strings.ToLower(k) {
+		case replLabel:
+			repl, err := strconv.Atoi(v)
+			if err != nil {
+				return "", 0, nil, fmt.Errorf("invalid replication factor %q, must be [1..3]", v)
+			}
+			spec.HaLevel = int64(repl)
+		case fsLabel:
+			fsType, err := osdapi.FSTypeSimpleValueOf(v)
+			if err != nil {
+				return "", 0, nil, fmt.Errorf("invalid fs value %q, must be ext4|btrfs", v)
+			}
+			spec.Format = fsType
+		case snapshotIntervalLabel:
+			si, err := strconv.Atoi(v)
+			if err != nil {
+				return "", 0, nil, fmt.Errorf("invalid snapshot interval value %q, must be a number", v)
+			}
+			err = checkSnapInterval(si)
+			if err != nil {
+				return "", 0, nil, err
+			}
+			spec.SnapshotInterval = uint32(si)
+		case cosLabel:
+			cos, err := cosLevel(v)
+			if err != nil {
+				return "", 0, nil, err
+			}
+			spec.Cos = cos
+		case sharedLabel:
+			spec.Shared = true
+		case blocksizeLabel:
+			blocksize, err := strconv.Atoi(v)
+			if err != nil {
+				return "", 0, nil, fmt.Errorf("invalid block_size value %q, must be a number ", v)
+			}
+			spec.BlockSize = int64(blocksize)
+		case aggregationLabel:
+			agg, err := strconv.Atoi(v)
+			if err != nil {
+				return "", 0, nil, fmt.Errorf("invalid aggregation_level %q, must be a number ", v)
+			}
+			spec.AggregationLevel = uint32(agg)
+		default:
+			labels[strings.ToLower(k)] = v
+		}
+	}
+	source := osdapi.Source{}
```

Comment:
I tried to keep it same with other volume drivers here. All of them do some kind of parsing here. Either we do it here or we pass it to openstorage and error back.